### PR TITLE
feat: support for reduce

### PR DESCRIPTION
**Pull request by yhl25**
Signed-off-by: Yashash H L <yashashhl25@gmail.com>

adding support for reduce, minor code refactoring and formatting

### CODE REVIEW

**Comment by vigith** [comment]
looks like this includes a reformatting of indentation, was this expected?

**Comment by yhl25** [comment]
> looks like this includes a reformatting of indentation, was this expected? if yes, should we standardize on the indentation format for Java? e.g, Golang has `gofmt`, Python SDK uses `black`.

Yes, I use IntelliJ idea default format

File: examples/src/main/java/io/numaproj/numaflow/examples/function/sum/SumFunction.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        while (true) {
+            Udfunction.Datum datum = reduceDatumStream.ReadMessage();
+            // null indicates the end of the input
+            if (datum == null) {
+                break;
+            }
+            try {
+                sum += Integer.parseInt(new String(datum.getValue().toByteArray()));
+            } catch (NumberFormatException e) {
+                System.out.println("unable to convert the value to int, " + e.getMessage());
```
Comment (ashwinidulams) [inline examples/src/main/java/io/numaproj/numaflow/examples/function/sum/SumFunction.java:30]: remove sop's

File: src/main/java/io/numaproj/numaflow/function/FunctionServer.java [hunk 2]
```diff
@@ -15,100 +17,107 @@
...
+        // register shutdown hook
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+            // Use stderr here since the logger may have been reset by its JVM shutdown hook.
+            System.err.println("*** shutting down gRPC server since JVM is shutting down");
+            try {
+                FunctionServer.this.stop();
+            } catch (InterruptedException e) {
+                e.printStackTrace(System.err);
+            }
+            System.err.println("*** server shut down");
```
Comment (ashwinidulams) [inline src/main/java/io/numaproj/numaflow/function/FunctionServer.java:110]: here

File: src/main/java/io/numaproj/numaflow/function/FunctionService.java [hunk 3]
```diff
@@ -3,64 +3,165 @@
...
+        Metadata md = new MetadataImpl(iw);
+
 
-    // get Datum from request
-    Udfunction.Datum handlerDatum = Udfunction.Datum.newBuilder()
-        .setValue(request.getValue())
-        .setEventTime(request.getEventTime())
-        .setWatermark(request.getWatermark())
-        .build();
+        ExecutorService executor = Executors.newSingleThreadExecutor();
```
Comment (ashwinidulams) [inline src/main/java/io/numaproj/numaflow/function/FunctionService.java:101]: consider adding a shutdown hook that can cancel the tasks. 

Also consider using simple Callable with FutureTask. although the incidence of too many executors is rare, we have seen deadlocks with too many executors.

File: examples/src/main/java/io/numaproj/numaflow/examples/function/sum/SumFunction.java (hunk 1 above)
Comment (yhl25) [reply examples/src/main/java/io/numaproj/numaflow/examples/function/sum/SumFunction.java]: done

File: src/main/java/io/numaproj/numaflow/function/FunctionServer.java (hunk 2 above)
Comment (yhl25) [reply src/main/java/io/numaproj/numaflow/function/FunctionServer.java:122]: done

File: src/main/java/io/numaproj/numaflow/function/FunctionService.java (hunk 3 above)
Comment (yhl25) [reply src/main/java/io/numaproj/numaflow/function/FunctionService.java]: added fixed thread pool

**Review by ashwinidulams — APPROVED** [review]

File: .editorconfig [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+# ij_java_field_annotation_wrap = split_into_lines
+# ij_java_finally_on_new_line = false
+# ij_java_for_brace_force = never
+# ij_java_for_statement_new_line_after_left_paren = false
+# ij_java_for_statement_right_paren_on_new_line = false
+# ij_java_for_statement_wrap = off
+# ij_java_generate_final_locals = false
+# ij_java_generate_final_parameters = false
+# ij_java_if_brace_force = never
+ij_java_imports_layout = org.apache.flink.**, |, org.apache.flink.shaded.**, |, *, |, javax.**, |, java.**, |, scala.**, |, $*
```
Comment (vigith) [inline .editorconfig:116]: please clean up editorconfig, remove the flink references and also the commented out ones.

**Pull request merged**